File: Leetcode/Array/Search in Rotated Sorted Array.py
# O(log N)  O(1)
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if not nums: return -1
        else:
            start = 0
            end = len(nums)-1
            
            while (start <= end):
                mid = (start+end)//2
                
                # If the mid element is the target, return index
                if nums[mid] == target:
                    return mid
                # If not, need to determine the rotation point of the array before dividing up to search
                else:
                    # If the array is rotated in the later half
                    if nums[start] <= nums[mid]:
                        if target >= nums[start] and target < nums[mid]:
                            end = mid - 1
                        else:
                            start = mid + 1
                    # If the array is rotated in the previous half
                    elif nums[mid] < nums[end]:
                        if target > nums[mid] and target <= nums[end]:
                            start = mid + 1
                        else:
                            end = mid - 1
            return -1
       

File: Java/Search in Rotated Sorted Array.java

//1. Find a rotation index rotation_index, i.e. index of the smallest element in the array. 
//2. rotation_index splits array in two parts. 
//3. Compare nums[0] and target to identify in which part one has to look for target.
//4.Perform a binary search in the chosen part of the array.

public class Solution {
    public int search(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        int start = 0;
        int end = nums.length - 1;
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < nums[end]) {//if nums[mid] < nums[end], we check whether target is on the right of mid
                if (target > nums[mid] && target <= nums[end]) {//this part is sorted,check whether target is in this range
                    start = mid;
                } else {
                    end = mid;
                }
            } else {//if nums[mid] > nums[end], we check whether target is on the left of mid
                if (target < nums[mid] && target >= nums[start]) {//this part is sorted,check whether target is in this range
                    end = mid;
                } else {
                    start = mid;
                }
            }
        }
        if (nums[start] == target) {
            return start;
        }
        if (nums[end] == target) {
            return end;
        }
        return -1;
    }
}










